### PR TITLE
optee/rpmb.c: Fix driver routing memory alignment using a temporary buffer

### PR DESCRIPTION
**Pull request by litchipi**
As optee-os append a header of size 6 before a raw RPMB frame, it creates
an alignment error in the rpmb driver.
To prevent breaking ABI of OPTEE-OS RPC memrefs, allocate a temporary
buffer to copy the data into an aligned memory.

Signed-off-by: Timothée Cercueil <timothee.cercueil@st.com>

### CODE REVIEW

File: drivers/tee/optee/rpmb.c [hunk 1]
```diff
@@ -131,8 +141,12 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
 		mmc = get_mmc(priv, sreq->dev_id);
 		if (!mmc)
 			return TEE_ERROR_ITEM_NOT_FOUND;
-		if (mmc_rpmb_route_frames(mmc, RPMB_REQ_DATA(req),
-					  req_size - sizeof(struct rpmb_req),
+
+		if (!rpmb_frame_buffer)
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]: i would remove the change at line 72 and add here:
```
			if (!rpmb_frame_buffer) {
				rpmb_frame_buffer = malloc(RPMB_FRAME_BUFFER_SIZE);
				if (!rpmb_frame_buffer)
					return TEE_ERROR_OUT_OF_MEMORY;
			}
			memcpy(rpmb_frame_buffer, RPMB_REQ_DATA(req), RPMB_FRAME_SIZE);
```

Also, here we're not sure req_size == RPMB_FRAME_BUFFER_SIZE + sizeof(sreq);
If it differs from, then we must allocation/release the RPMB temporary buffer at each call :(

Comment (litchipi) [reply drivers/tee/optee/rpmb.c]: I would prefer a persistence of the buffer, this way a new buffer will be allocated only if it needs more space. In any case the buffer is bonded by the request size so wont be bothered by previous data, and can be checked for free at release of mmc.
Something like this for exemple:
``` C
static size_t rpmb_frame_buffer_sz = 512;
static void* rpmb_frame_buffer;

release_mmc()
{
    if (rpmb_frame_buffer)
        free(rpmb_frame_buffer);
}

rpmb_cmd()
{
    rpmb_frame_data_sz = (req_size - sizeof(struct rpmb_req));
    if ((!rpmb_frame_buffer) || (rpmb_frame_data_sz > rpmb_frame_buffer_sz))
    {
        if (rpmb_frame_buffer)
            free(rpmb_frame_buffer);
        rpmb_frame_buffer_sz = rpmb_frame_data_sz;
        rpmb_frame_buffer = malloc(rpmb_frame_buffer_sz);
    }
    memcpy( ... )
    rpmb_route_frame( ... )
}
```

**Comment by litchipi** [comment]
I implemented the solution using a dynamic buffer size (starting at 512) that we simply re-allocate if more space is required at some point.
It ensure that the buffer is always allocated before routing frame, that the malloc succeeds and that the buffer is freed on mmc release.

File: drivers/tee/optee/rpmb.c [hunk 2]
```diff
@@ -131,8 +142,23 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
...
+				free(rpmb_frame_buffer);
+
+			if (rpmb_frame_data_sz > rpmb_frame_buffer_sz)
+				rpmb_frame_buffer_sz = rpmb_frame_data_sz;
+
+			rpmb_frame_buffer = malloc(rpmb_frame_buffer_sz);
+			if (!rpmb_frame_buffer)
+				return TEE_ERROR_OUT_OF_MEMORY;
+		}
+		memcpy(rpmb_frame_buffer, RPMB_REQ_DATA(req), rpmb_frame_data_sz);
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]: this allocates and copy bffer even if rpmb frame buffer is 32bit aligned. In such case, we could pass it straight to `mmc_rpmb_route_frames()`.

Comment (litchipi) [reply drivers/tee/optee/rpmb.c]: I fixed this, but it adds a big block in a previously seamless code, would it be worth it to create another function to get the aligned address ?
Which would return `RPMB_REQ_DATA(req)` if the alignment is correct, and if not creates the buffer if needed and copy the data in it.
It would return a void* to the address of either the buffer or the data after the request
Something like `void* get_aligned_rpmb_data(void* addr, size_t data_len)`

**Comment by litchipi** [comment]
> this allocates and copy bffer even if rpmb frame buffer is 32bit aligned. In such case, we could pass it straight to `mmc_rpmb_route_frames()`.

I fixed this, but it adds a big block in a previously seamless code, would it be worth it to create another function to get the aligned address ?
Which would return `RPMB_REQ_DATA(req)` if the alignment is correct, and if not creates the buffer if needed and copy the data in it.
It would return a void* to the address of either the buffer or the data after the request
Something like `void* get_aligned_rpmb_data(void* addr, size_t data_len)`

**Comment by litchipi** [comment]
(squashed previous changes, removed from blank lines)

File: drivers/tee/optee/rpmb.c [hunk 3]
```diff
@@ -131,8 +141,29 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
 		mmc = get_mmc(priv, sreq->dev_id);
 		if (!mmc)
 			return TEE_ERROR_ITEM_NOT_FOUND;
-		if (mmc_rpmb_route_frames(mmc, RPMB_REQ_DATA(req),
-					  req_size - sizeof(struct rpmb_req),
+
+		rpmb_frame_data_sz = (req_size - sizeof(struct rpmb_req));
+		if (((unsigned long) RPMB_REQ_DATA(req)) & 0x03) // Unaligned
+		{
+			if ((!rpmb_frame_buffer) || (rpmb_frame_data_sz > rpmb_frame_buffer_sz))
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]: `if (!rpmb_frame_buffer || (rpmb_frame_data_sz > rpmb_frame_buffer_sz))`

File: drivers/tee/optee/rpmb.c [hunk 4]
```diff
@@ -131,8 +141,29 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
 		mmc = get_mmc(priv, sreq->dev_id);
 		if (!mmc)
 			return TEE_ERROR_ITEM_NOT_FOUND;
-		if (mmc_rpmb_route_frames(mmc, RPMB_REQ_DATA(req),
-					  req_size - sizeof(struct rpmb_req),
+
+		rpmb_frame_data_sz = (req_size - sizeof(struct rpmb_req));
+		if (((unsigned long) RPMB_REQ_DATA(req)) & 0x03) // Unaligned
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]: no space between cast and casted value:
prefer remove `// some comment` at end of line.

`		if ((uintptr_t)RPMB_REQ_DATA(req) & 0x03)`

**Review by etienne-lms — COMMENTED** [review]
In commit log:

`As optee-os append a header of size 6 before a raw RPMB frame, it creates`

Replace `append` by `insert`: the header is before the payload we need to align.
Replace `header of size 6` with `6-byte header`.

s/optee-os/OP-TEE OS/
s/rpmb/RPMB/

**Comment by litchipi** [comment]
I fixed all that, thanks

**Comment by litchipi** [comment]
Pushed changes after second review, removed globals as it's badly seen by U-Boot community.
Simpler solution overall

File: drivers/tee/optee/rpmb.c [hunk 5]
```diff
@@ -131,10 +136,24 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
...
+			memcpy(rpmb_frame_buffer, RPMB_REQ_DATA(req), rpmb_frame_data_sz);
+			rpmb_data = rpmb_frame_buffer;
+		}
+		else
+			rpmb_data = RPMB_REQ_DATA(req);
+		ret = mmc_rpmb_route_frames(mmc, rpmb_data, rpmb_frame_data_sz, rsp, rsp_size);
+		if (rpmb_frame_buffer)
+		   free(rpmb_frame_buffer);
+		if (ret)
+		  return TEE_ERROR_BAD_PARAMETERS;
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]: bad indentation: use tabulation.

File: drivers/tee/optee/rpmb.c [hunk 6]
```diff
@@ -131,10 +136,24 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
...
+			/* data alignment is required by RPMB */
+			rpmb_frame_buffer = malloc(rpmb_frame_data_sz);
+			if (!rpmb_frame_buffer)
+				return TEE_ERROR_OUT_OF_MEMORY;
+			memcpy(rpmb_frame_buffer, RPMB_REQ_DATA(req), rpmb_frame_data_sz);
+			rpmb_data = rpmb_frame_buffer;
+		}
+		else
+			rpmb_data = RPMB_REQ_DATA(req);
+		ret = mmc_rpmb_route_frames(mmc, rpmb_data, rpmb_frame_data_sz, rsp, rsp_size);
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]: Maybe prefer:
```
		rpmb_frame_data_sz = (req_size - sizeof(struct rpmb_req));
		rpmb_data = RPMB_REQ_DATA(req);

		if ((uintptr_t)rpmb_data & 0x03)
		{
			/* Data alignment is required by RPMB */
			rpmb_frame_buffer = malloc(rpmb_frame_data_sz);
			if (!rpmb_frame_buffer)
				return TEE_ERROR_OUT_OF_MEMORY;
			memcpy(rpmb_frame_buffer, RPMB_REQ_DATA(req), rpmb_frame_data_sz);
			rpmb_data = rpmb_frame_buffer;
		}

		ret = mmc_rpmb_route_frames(mmc, rpmb_data, rpmb_frame_data_sz, rsp, rsp_size);

		if (rpmb_frame_buffer)
			free(rpmb_frame_buffer);
		if (ret)
			return TEE_ERROR_BAD_PARAMETERS;
		return TEE_SUCCESS;

**Review by etienne-lms — CHANGES_REQUESTED** [review]

**Comment by litchipi** [comment]
Updated log message, added examples of drivers affected by the problem this patch resolves

File: drivers/tee/optee/rpmb.c [hunk 7]
```diff
@@ -122,6 +122,11 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
 {
 	struct rpmb_req *sreq = req;
 	struct mmc *mmc;
+	size_t rpmb_frame_data_sz;
+	void* rpmb_data;
+	void* rpmb_frame_buffer = NULL;
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]:
```patch
-	void* rpmb_data;
-	void* rpmb_frame_buffer = NULL;
+	void *rpmb_data;
+	void *rpmb_frame_buffer = NULL;
```

File: drivers/tee/optee/rpmb.c (hunk 5 above)
Comment (etienne-lms) [reply drivers/tee/optee/rpmb.c]: fix indentation

File: drivers/tee/optee/rpmb.c [hunk 8]
```diff
@@ -131,10 +136,24 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
...
+			rpmb_frame_buffer = malloc(rpmb_frame_data_sz);
+			if (!rpmb_frame_buffer)
+				return TEE_ERROR_OUT_OF_MEMORY;
+			memcpy(rpmb_frame_buffer, RPMB_REQ_DATA(req), rpmb_frame_data_sz);
+			rpmb_data = rpmb_frame_buffer;
+		}
+
+		ret = mmc_rpmb_route_frames(mmc, rpmb_data, rpmb_frame_data_sz, rsp, rsp_size);
+		if (rpmb_frame_buffer)
+		   free(rpmb_frame_buffer);
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]: can use `free(rpmb_frame_buffer);` straight, no need to test `rpmb_frame_buffer != NULL`. 
`free(NULL);` is a valid call.

File: drivers/tee/optee/rpmb.c [hunk 9]
```diff
@@ -131,10 +136,24 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
...
 		if (!mmc)
 			return TEE_ERROR_ITEM_NOT_FOUND;
-		if (mmc_rpmb_route_frames(mmc, RPMB_REQ_DATA(req),
-					  req_size - sizeof(struct rpmb_req),
-					  rsp, rsp_size))
-			return TEE_ERROR_BAD_PARAMETERS;
+
+		rpmb_data = RPMB_REQ_DATA(req);
+		rpmb_frame_data_sz = (req_size - sizeof(struct rpmb_req));
+		if ((uintptr_t)RPMB_REQ_DATA(req) & 0x03)
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]: the address value is already available in `rpmb_data`, prefer use it for the test:
```patch
-		if ((uintptr_t)RPMB_REQ_DATA(req) & 0x03)
+		if ((uintptr_t)rpmb_data & 0x03)
```
Ditto at line 148.

File: drivers/tee/optee/rpmb.c [hunk 10]
```diff
@@ -131,10 +136,24 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
 		mmc = get_mmc(priv, sreq->dev_id);
 		if (!mmc)
 			return TEE_ERROR_ITEM_NOT_FOUND;
-		if (mmc_rpmb_route_frames(mmc, RPMB_REQ_DATA(req),
-					  req_size - sizeof(struct rpmb_req),
-					  rsp, rsp_size))
-			return TEE_ERROR_BAD_PARAMETERS;
+
+		rpmb_data = RPMB_REQ_DATA(req);
+		rpmb_frame_data_sz = (req_size - sizeof(struct rpmb_req));
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]: outer parentheses not need.

File: drivers/tee/optee/rpmb.c [hunk 11]
```diff
@@ -131,10 +136,24 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
...
+		{
+			/* data alignment is required by RPMB */
+			rpmb_frame_buffer = malloc(rpmb_frame_data_sz);
+			if (!rpmb_frame_buffer)
+				return TEE_ERROR_OUT_OF_MEMORY;
+			memcpy(rpmb_frame_buffer, RPMB_REQ_DATA(req), rpmb_frame_data_sz);
+			rpmb_data = rpmb_frame_buffer;
+		}
+
+		ret = mmc_rpmb_route_frames(mmc, rpmb_data, rpmb_frame_data_sz, rsp, rsp_size);
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]: Warning: 80 character per line max.

File: drivers/tee/optee/rpmb.c [hunk 12]
```diff
@@ -122,6 +122,11 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
 {
 	struct rpmb_req *sreq = req;
 	struct mmc *mmc;
+	size_t rpmb_frame_data_sz;
+	void* rpmb_data;
+	void* rpmb_frame_buffer = NULL;
+	u32 ret;
+
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]: remove this extra empty line.

File: drivers/tee/optee/rpmb.c [hunk 13]
```diff
@@ -122,6 +122,11 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
 {
 	struct rpmb_req *sreq = req;
 	struct mmc *mmc;
+	size_t rpmb_frame_data_sz;
+	void* rpmb_data;
+	void* rpmb_frame_buffer = NULL;
+	u32 ret;
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]: Should be an `int`, see `mmc_rpmb_route_frames()` return value.

**Comment by litchipi** [comment]
Patch fixed and comitted. I removed the Change-Id in the footer of the commit message before pushing as I thought it wasn't necessary

File: drivers/tee/optee/rpmb.c [hunk 14]
```diff
@@ -131,9 +135,23 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
...
 		if (!mmc)
 			return TEE_ERROR_ITEM_NOT_FOUND;
-		if (mmc_rpmb_route_frames(mmc, RPMB_REQ_DATA(req),
-					  req_size - sizeof(struct rpmb_req),
-					  rsp, rsp_size))
+
+		rpmb_data = RPMB_REQ_DATA(req);
+		rpmb_frame_data_sz = req_size - sizeof(struct rpmb_req);
+		if ((uintptr_t)rpmb_data & 0x03)
+		{
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]: U-Boot coding style expects:
```c
		if ((uintptr_t)rpmb_data & 0x03) {
			(...)
		}

File: drivers/tee/optee/rpmb.c [hunk 15]
```diff
@@ -131,9 +135,23 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
...
+			/* data alignment is required by RPMB */
+			rpmb_frame_buffer = malloc(rpmb_frame_data_sz);
+			if (!rpmb_frame_buffer)
+				return TEE_ERROR_OUT_OF_MEMORY;
+			memcpy(rpmb_frame_buffer, rpmb_data, rpmb_frame_data_sz);
+			rpmb_data = rpmb_frame_buffer;
+		}
+
+		ret = mmc_rpmb_route_frames(mmc, rpmb_data, rpmb_frame_data_sz,
+				rsp, rsp_size);
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]: indent to parenthesis (add 1 tab and 4 space chars):
```c
		ret = mmc_rpmb_route_frames(mmc, rpmb_data, rpmb_frame_data_sz,
					    rsp, rsp_size);

File: drivers/tee/optee/rpmb.c [hunk 16]
```diff
@@ -122,6 +122,10 @@ static u32 rpmb_process_request(struct optee_private *priv, void *req,
 {
 	struct rpmb_req *sreq = req;
 	struct mmc *mmc;
+	size_t rpmb_frame_data_sz;
```
Comment (etienne-lms) [inline drivers/tee/optee/rpmb.c]: minor: for consistency I would rename this `rpmb_data_sz` and move this line below line 127.

**Comment by litchipi** [comment]
Fixed everything (pushed new commits instead of force-push amend, will squash everything before sending to mailing list)

**Comment by litchipi** [comment]
Closing PR as the patch review is over
